compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5189,6 +5191,262 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+
+   const uint8_t len = isLong ? 8 : 4;
+
+   generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldarx : TR::InstOpCode::lwarx, node, oldValueReg,
+         new (cg->trHeapMemory()) TR::MemoryReference(0, valueAddrReg, len, cg));
+
+   if (isAddOp)
+      {
+      numDeps++;
+      newValueReg = cg->allocateRegister();
+
+      if (isDeltaImmediateShifted)
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addis, node,
+               newValueReg, oldValueReg, ((delta & 0xFFFF0000) >> 16));
+      else if (isDeltaImmediate)
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node,
+               newValueReg, oldValueReg, delta);
+      else
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, newValueReg,
+               oldValueReg, deltaReg);
+      }
+   else
+      {
+      newValueReg = deltaReg;
+      deltaReg = NULL;
+      }
+
+   generateMemSrc1Instruction(cg, isLong ? TR::InstOpCode::stdcx_r : TR::InstOpCode::stwcx_r, node, new (cg->trHeapMemory()) TR::MemoryReference(0, valueAddrReg, len, cg),
+         newValueReg);
+
+   // We expect this store is usually successful, i.e., the following branch will not be taken
+   if (TR::Compiler->target.cpu.id() >= TR_PPCgp)
+      {
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, PPCOpProp_BranchUnlikely, node, loopLabel, cndReg);
+      }
+   else
+      {
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, loopLabel, cndReg);
+      }
+
+   // We deviate from the VM helper here: no-store-no-barrier instead of always-barrier
+   generateInstruction(cg, TR::InstOpCode::sync, node);
+
+   TR::RegisterDependencyConditions *conditions;
+
+   //Set the conditions and dependencies
+   conditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions((uint16_t) numDeps, (uint16_t) numDeps, cg->trMemory());
+
+   addDependency(conditions, valueAddrReg, TR::RealRegister::NoReg, TR_GPR, cg);
+   addDependency(conditions, oldValueReg, TR::RealRegister::NoReg, TR_GPR, cg);
+   conditions->getPreConditions()->getRegisterDependency(1)->setExcludeGPR0();
+   conditions->getPostConditions()->getRegisterDependency(1)->setExcludeGPR0();
+   addDependency(conditions, cndReg, TR::RealRegister::cr0, TR_CCR, cg);
+   numDeps = numDeps - 3;
+
+   if (newValueReg)
+      {
+      addDependency(conditions, newValueReg, TR::RealRegister::NoReg, TR_GPR, cg);
+      numDeps--;
+      }
+
+   if (deltaReg)
+      {
+      addDependency(conditions, deltaReg, TR::RealRegister::NoReg, TR_GPR, cg);
+      numDeps--;
+      }
+
+   doneLabel->setEndInternalControlFlow();
+   generateDepLabelInstruction(cg, TR::InstOpCode::label, node, doneLabel, conditions);
+
+   cg->decReferenceCount(valueAddrChild);
+   cg->stopUsingRegister(valueAddrReg);
+   cg->stopUsingRegister(cndReg);
+
+   if (deltaReg)
+      {
+      cg->stopUsingRegister(deltaReg);
+      }
+
+   if (deltaChild)
+      {
+      cg->decReferenceCount(deltaChild);
```

Comment:
Furthermore, deltaReg might be returned as resultReg.  You had better setRegister before this decReferenceCount, because it can prematurely end that register's life (if this were the last reference to deltaChild).

Reply:
Fixed in https://github.com/eclipse/omr/pull/3018/commits/a453a47d5b03cd028f0c9be8e0ed55b752ea2d70